rfcs/017 - Cluster Data Storage RFC.md

```diff
@@ -0,0 +1,184 @@
+
+But those endpoints make clear something else that's necessary: the `HandlerMemory` structs that represent the value of a particular key need to be able to be serialized and deserialized, and in a way that they can be done so not just in Rust, but any language we want to fully support in Anycloud. (That's not entirely correct, we could create a set of secondary endpoints just for Anycloud that only accepts strings, or rather `Array<int8>` arrays where strings are a special case of that, but that may limit/impede future efforts at RPC between the Alan and Anycloud for offloading CPU-intensive work to Alan, and we still need serialization/deserialization of HandlerMemory for Alan itself.)
+
+Since all nodes in a cluster will be running the same version of Alan, which means they all have the same HandlerMemory, [protobuf](https://github.com/protocolbuffers/protobuf) is the best solution here for speed, size, and compatibility concerns, as it officially supports many languages and there is also a [Rust implementation](https://docs.rs/protobuf/2.22.1/protobuf/) that can be used. There are other protocols that are more widespread (JSON), faster (CapnProto), and smaller (Serde), but Protobuf hits a sweet spot where it's decent at all three (and it isn't an Apache project monstrosity, will veto any push for something out of apache.org).
+
+### Global Mutable State that Actually Works
+
+In many cases where `@std/datastore` is simply used as a cache, we are done, but in any case where one or more keys need to have their value mutated, not just replaced, we need a way to atomically perform that mutation so we're sure that other places where a similar mutation could be occurring simultaneously don't accidentally obliterate the work we are doing.
+
+To be clear, this vulnerability affects `@std/datastore` as it currently exists even without the state being distributed across a cluster. It's just harder to hit when the latencies are lower when completely local. The traditional approach of locks on shared mutable state, besides being complicated to follow for many developers, also becomes a performance bottleneck when the locks are distributed across a cluster and lock acquisition and release times are several orders of magnitude slower. For this reason, an approach based on Software Transactional Memory (STM) is the choice to handle mutation of one-or-more datastore keys.
+
+Because STM re-runs the code involved in a mutation of memory, side-effects can be triggered multiple times. If one of those "side-effects" is billing a customer through Stripe, for instance, that would be a *very* bad thing to do automatically without any indication to the user! Also because the STM behavior increases the total latency to perform a mutation (as it needs to confirm that all writes are good before committing them, more on that algo later) being able to escape that behavior when using datastore as just a cache is also desireable.
```

Comment:
Side note: I wonder if it's possible to fake atomicity in our http implementation to ensure that the http request would only be sent 1x in this case?

Reply:
Unfortunately not really possible.

1. If the HTTP request is to a static URL with a static payload, then in only that case *maybe* (see point 2 for why even then sometimes no), but if that URL and/or payload depend on something constructed during the `atomic` operation each one would be different and it would be hard to tell which is right.
2. If you try to delay calling the HTTP request until after the `atomic` execution has succeeded, you could cause the equivalent of a deadlock if that HTTP *response* is part of the data to be put into the datastore, so the STM mode is waiting for data that it doesn't have because its waiting for STM mode to finish, therefore it can never finish.

So because of those two points, silently running STM all the time isn't something that could work "correctly" from the perspective of a developer used to single-threaded imperative programming. That's why explicitly calling it out in a function is worthwhile.